x-pack/plugins/fleet/server/services/agents/build_status_runtime_field.ts

```diff
@@ -0,0 +1,116 @@
+
+import { AGENT_POLLING_THRESHOLD_MS } from '../../constants';
+import { agentPolicyService } from '../agent_policy';
+const DEFAULT_MS_BEFORE_INACTIVE = 10 * 60 * 1000; // 10 minutes
+const MISSED_INTERVALS_BEFORE_OFFLINE = 10;
+const MS_BEFORE_OFFLINE = MISSED_INTERVALS_BEFORE_OFFLINE * AGENT_POLLING_THRESHOLD_MS;
+
+export type InactivityTimeouts = Awaited<
+  ReturnType<typeof agentPolicyService['getInactivityTimeouts']>
+>;
+
+const _buildInactiveClause = (
+  now: number,
+  inactivityTimeouts: InactivityTimeouts,
+  field: (path: string) => string
+) => {
+  const policyClauses = inactivityTimeouts
+    .map(({ inactivityTimeout, policyIds }) => {
+      const inactivityTimeoutMs = inactivityTimeout * 1000;
+      const policyOrs = policyIds
+        .map((policyId) => `${field('policy_id')}.value == '${policyId}'`)
+        .join(' || ');
+
+      return `(${policyOrs}) && lastCheckinMillis < ${now - inactivityTimeoutMs}L`;
+    })
+    .join(' || ');
+
+  const policyHasNoInactivityTimeout = `lastCheckinMillis < ${now - DEFAULT_MS_BEFORE_INACTIVE}L`;
+  const agentIsInactive =
+    (policyClauses.length ? `(${policyClauses}) || ` : '') + policyHasNoInactivityTimeout;
+  return `lastCheckinMillis > 0 && ${field('policy_id')}.size() > 0 && (${agentIsInactive})`;
+};
+
+function _buildSource(inactivityTimeouts: InactivityTimeouts, pathPrefix?: string) {
+  const normalizedPrefix = pathPrefix ? `${pathPrefix}${pathPrefix.endsWith('.') ? '' : '.'}` : '';
+  const field = (path: string) => `doc['${normalizedPrefix + path}']`;
+  const now = Date.now();
+  return `
+    long lastCheckinMillis = ${field('last_checkin')}.size() > 0 
+      ? ${field('last_checkin')}.value.toInstant().toEpochMilli() 
+      : -1;
+    if (${field('unenrolled_at')}.size() > 0) {
+      emit('unenrolled'); 
+    } else if (${_buildInactiveClause(now, inactivityTimeouts, field)}) {
+      emit('inactive');
+    } else if (
+        lastCheckinMillis > 0 
+        && lastCheckinMillis 
+        < (${now - MS_BEFORE_OFFLINE}L)
+    ) { 
+      emit('offline'); 
+    } else if (
+      ${field('policy_revision_idx')}.size() == 0 || (
+        ${field('upgrade_started_at')}.size() > 0 &&
+        ${field('upgraded_at')}.size() == 0
+      )
+    ) { 
+      emit('updating'); 
+    } else if (${field('last_checkin')}.size() == 0) {
+      emit('enrolling'); 
+    } else if (${field('unenrollment_started_at')}.size() > 0) {
+      emit('unenrolling'); 
+    } else if (
+      ${field('last_checkin_status')}.size() > 0 &&
+      ${field('last_checkin_status')}.value.toLowerCase() == 'error'
+    ) { 
+        emit('error');
+    } else if (
+      ${field('last_checkin_status')}.size() > 0 &&
+      ${field('last_checkin_status')}.value.toLowerCase() == 'degraded'
+    ) { 
+      emit('degraded');
+    } else { 
+      emit('online'); 
+    }
+    `.replace(/(\n|\s{4})/g, ''); // condense source to single line
```

Comment:
I think the script works with multiple lines, I have used like that in update tags.

Reply:
Yep it does 👍